### DISCUSSION TITLE
Creating a 3D mesh projecting a surface

### DISCUSSION THREAD

**Discussion by E1k3** [jimy-byerley/pymadcad#15] [Q&A] [answered]
Hi, I was reading parts of the pymadcad code to learn more about 3D surface offsetting and I am currently facing a problem which you might have an answer to.
I am trying to create a 3D body by projecting a surface to Z=0, i.e. I use the original surface as the top, the flipped projection as the bottom and connect both with a junction made of quads. The problem is, that the original surface may have folds ([like this](https://i.imgur.com/V8gDCO5.png)), which results in self intersections in the projection as well as the junction ([like this](https://imgur.com/fZYpHJO)).

Is there an elegant way to achieve this or do I have to iterate through the mesh and fix each intersection individually?

**Comment by jimy-byerley** [comment, marked as answer]
Hello @Etk3 and welcome here !
If I correctly understood, you are looking to a way to achieve this:
![image](https://user-images.githubusercontent.com/7940753/148590932-b2a4bcb1-33df-4cfe-820d-0ce2c7e057fb.png)

The good new is that there is few ways to acheive it. None are as simple as en extrusion ...
```python
curve = wire(Softened([
		vec3(-0.2296, 0, 0.1898),
		vec3(-0.08743, 0, 0.1648),
		vec3(-0.1569, 0, 0.06546),
		vec3(0.2357, 0, 0.08456),
		vec3(0.1315, 0, 0.1761),
		vec3(0.1717, 0, 0.2101),
		vec3(0.3164, 0, 0.1736)]))
surf = extrusion(0.1*Y, curve)

mesh_extruded = extrusion(mat3(1,1,-1), surf)    # the mat3 here makes a symetry in direction Z
```
This is giving what you already have
![image](https://user-images.githubusercontent.com/7940753/148591277-4f968684-329e-4d66-8aba-ea1835859fea.png)

1. The most efficient way, if you know the surface side curves by advance, and if your side surfaces are planar is to generate the side surfaces by triangulation (well the side surfaces can be not perfectly planar, but must be quite regular)

  ```python
  mesh_generated = (	surf   # original top surface
                   +	surf.transform(mat3(1,1,-1)).flip()  # generate a symetric lower surface with the symetry transformation
                   +	flatsurface(curve.flip() + curve.transform(mat3(1,1,-1)))  # generate a triangulated surface from generated symetric curves
                   )
  ```
  ![image](https://user-images.githubusercontent.com/7940753/148592381-197cbd4a-59a1-4cb9-8815-3e93ddd8d1ef.png)

2. If you do not already have the side curves, you can get it by propagative selection (that's an undocumented feature of madcad)
  ```python
  from madcad.selection import select, edgenear, stopangle
  
  lower = surf.transform(mat3(1,1,-1))  # generate the lower surface with a symetry transformation
  some = (	wire(select(surf, edgenear(surf.outlines(), Y), stopangle(0.5)))   # select the surface edges that are the closest to point Y and then select until there is an angle above 0.5 rad
	  + 	wire(select(lower, edgenear(lower.outlines(), Y), stopangle(0.5))).flip() )  # same selection for the lower surface
  mesh_triangulated = flatsurface(some)
  ```
  (same result as before)

3. If you want to have a specific kind of side surface, I would recommend to use the boolean operations. The inconvenient is that you must provide a surface that stops exactly on both curve ends. Or the boolean operation may fail to give a satisfactory result.
```python
# option 1:  flat side surface
side = flatsurface(wire([   # this is a simple rectangular plane
		vec3(curve.points[0].x, 0, 0.3),   # always use x min and max coordinates from the surface, so we are sure the boolean operation will work
		vec3(curve.points[-1].x, 0, 0.3),
		vec3(curve.points[-1].x, 0, -0.3),
		vec3(curve.points[0].x, 0, -0.3),
	]).transform(0.01*Y)),
# option 2: weird surface, for the example
side = saddle(
		Softened([
			vec3(curve.points[0].x, -0.02, 0.5),  # take the precise x min  location of the surface
			vec3(0.07548, -0.02745, 0.5092),
			vec3(0.1703, 0.02395, 0.5),
			vec3(curve.points[-1].x, -0.03, 0.5),  # take the precise x max location of the surface
			]),
		Softened([  # second path stays on the plane as its x=0
			vec3(0, -0.08676, 0.483),
			vec3(0, -0.04412, 0.2477),
			vec3(0, 0.007296, -0.04252),
			vec3(0, -0.1125, -0.4239),]),
		)

# quite readable operation once we have a good surface:
# execute a boolean operation to cut and titch meshes
mesh_cutted = difference(
	surf + surf.transform(mat3(1,1,-1)).flip(),
	side,
	)
```
![image](https://user-images.githubusercontent.com/7940753/148595781-d67368ef-7755-4be1-a993-ef4bc30c0b91.png)
![image](https://user-images.githubusercontent.com/7940753/148595631-380cd138-a230-4189-b790-c9dab8323176.png)


Unfortunately I do not know other good ways to acheive such. I hope these options can be useful to you, else let me now !
(anyway, it both case let me now ;) )

**Reply by jimy-byerley** [reply]
Also I just wanted to point out that you can copy paste pictures in your github messages if you don't want to use imgur ;)

**Reply by E1k3** [reply]
Thank you for this very detailed explanation. I am actually trying to do something like this:
![screenshot-2022-01-10_12:51:13-926852061](https://user-images.githubusercontent.com/18334289/148761779-583f72b8-797e-47f0-8b21-893acaea4cd1.png)
So instead of flipping the surface, the outline would just be projected onto a plane below the surface. I guess this could be achieved by replacing the lower surface with a plane with the same outlines or just cutting the result of your method in half, so this already was very helpful.

I am interested to see the implementation of `flatsurface`, so I will have a good read of your `triangulation_outline` function this evening.

**Reply by jimy-byerley** [reply]
Oh this algorithm in `triangulation_outline` is quite simple actually. It's only reducing a contour point by point using the priorities defined by a good heuristic
Tell me if you need any more help for generating your model, but for what you said, that's it :+1:
